File: Practica3/estimar-planta.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat

# ============================================================
# 1. CARGAR DATOS DESDE MATLAB
# ============================================================

datos = loadmat("mediciones_y_u.mat")

# Para ver qué variables contiene el archivo:
print(datos.keys())

y = np.squeeze(datos[0]).astype(float)
u = np.squeeze(datos[1]).astype(float)
t = np.squeeze(datos[2]).astype(float)

Ts = np.mean(np.diff(t))

print(f"Ts = {Ts:.6f} s")
print(f"Muestras de y: {len(y)}")
print(f"Muestras de u: {len(u)}")

# ============================================================
# 2. ARMAR LA REGRESION
#
# Modelo:
#
# y[k] = a1*y[k-1] + a2*y[k-2] + b0*u[k]
#
# Entonces:
#
# Y = X theta
#
# theta = [a1, a2, b0]^T
# ============================================================

Y = y[2:]

X = np.column_stack((
    y[1:-1],   # y[k-1]
    y[:-2],    # y[k-2]
    u[2:]      # u[k]
))


# ============================================================
# 3. MINIMOS CUADRADOS
# ============================================================

theta, residuals, rank, singular_values = np.linalg.lstsq(
    X, Y, rcond=None
)

a1, a2, b0 = theta

print("\n--- Modelo discreto estimado ---")
print(f"a1 = {a1}")
print(f"a2 = {a2}")
print(f"b0 = {b0}")

print(
    f"\ny[k] = ({a1:.6f}) y[k-1] "
    f"+ ({a2:.6f}) y[k-2] "
    f"+ ({b0:.6f}) u[k]"
)


# ============================================================
# 4. PASAR DE a1,a2,b0 A A,B,C
#
# Planta continua:
#
#            C
# G(s) = -----------
#         s² + As + B
#
# donde:
# A = p1 + p2
# B = p1*p2
# C = p1*p2*c
# ============================================================

D = -1 / a2

A = (a1 * D - 2) / Ts

B = (D - 1 - A * Ts) / Ts**2

C = b0 * D / Ts**2

print("\n--- Coeficientes continuos ---")
print(f"A = p1 + p2   = {A}")
print(f"B = p1*p2     = {B}")
print(f"C = p1*p2*c   = {C}")


# ============================================================
# 5. OBTENER p1, p2 y c
#
# p1 y p2 son las raices de:
#
# p² - A*p + B = 0
# ============================================================

discriminante = A**2 - 4*B

if discriminante < 0:
    print("\nATENCION:")
    print("El discriminante es negativo.")
    print("Los polos obtenidos no corresponden a dos p reales.")
    print("Revisar modelo, datos o discretizacion.")

p1 = (A + np.sqrt(complex(discriminante))) / 2
p2 = (A - np.sqrt(complex(discriminante))) / 2

c = C / B

print("\n--- Parametros de la planta ---")
print(f"p1 = {p1}")
print(f"p2 = {p2}")
print(f"c  = {c}")


# ============================================================
# 6. PREDICCION DE UN PASO
#
# Usamos los y medidos anteriores para predecir y[k].
# Esto sirve para verificar la regresion.
# ============================================================

y_pred = a1*y[1:-1] + a2*y[:-2] + b0*u[2:]


# ============================================================
# 7. GRAFICAR MEDICION VS PREDICCION
# ============================================================

plt.figure()

plt.plot(t[2:], y[2:], label="Medido")
plt.plot(t[2:], y_pred, "--", label="Predicción")

plt.xlabel("Tiempo [s]")
plt.ylabel("Ángulo")
plt.grid()
plt.legend()
plt.title("Salida medida vs modelo discreto")

plt.show()

File: Practica3/estimar-planta.m
%% =========================================================
%  IDENTIFICACION DE LA PLANTA
%
%  Modelo continuo supuesto:
%
%              p1*p2*c
%  G(s) = -------------------
%          (s+p1)(s+p2)
%
%  Aproximacion discreta utilizada:
%
%  y[k] = a1*y[k-1] + a2*y[k-2] + b0*u[k]
%
%% =========================================================

clear;
clc;
close all;


%% 1. CARGAR DATOS

load("mediciones_y_u.mat");

% Primero miramos qué cargó MATLAB
whos


%% 2. EXTRAER LAS SEÑALES
%
% Si al cargar el archivo aparece la variable "out",
% usamos:

y = double(out.d1(:));       % salida: ángulo IMU
u = double(out.d2(:));       % entrada
t = double(out.tout(:));     % tiempo


%% 3. VERIFICACIONES

fprintf("Cantidad de muestras y: %d\n", length(y));
fprintf("Cantidad de muestras u: %d\n", length(u));
fprintf("Cantidad de muestras t: %d\n", length(t));

if length(y) ~= length(u) || length(y) ~= length(t)
    error("u, y y t no tienen la misma cantidad de muestras");
end

% Período de muestreo
Ts = mean(diff(t));

fprintf("Ts = %.8f s\n", Ts);
fprintf("fs = %.2f Hz\n", 1/Ts);


%% 4. GRAFICAR LOS DATOS ORIGINALES

figure;

subplot(2,1,1);
plot(t, u);
grid on;
xlabel("Tiempo [s]");
ylabel("u[k]");
title("Entrada");

subplot(2,1,2);
plot(t, y);
grid on;
xlabel("Tiempo [s]");
ylabel("Ángulo");
title("Salida IMU");


%% =========================================================
% 5. ARMAR LA REGRESION
%
% Modelo:
%
% y[k] = a1*y[k-1] + a2*y[k-2] + b0*u[k]
%
% Lo escribimos:
%
% Y = X * theta
%
% donde:
%
% theta = [a1; a2; b0]
%
%% =========================================================

Y = y(3:end);

X = [ ...
    y(2:end-1), ...     % y[k-1]
    y(1:end-2), ...     % y[k-2]
    u(3:end)     ...    % u[k]
];


%% 6. MINIMOS CUADRADOS

% En MATLAB:
%
% theta = X \ Y
%
% resuelve el problema de mínimos cuadrados.

theta = X \ Y;

a1 = theta(1);
a2 = theta(2);
b0 = theta(3);

fprintf("\n===== MODELO DISCRETO =====\n");
fprintf("a1 = %.10f\n", a1);
fprintf("a2 = %.10f\n", a2);
fprintf("b0 = %.10f\n", b0);

fprintf("\ny[k] = %.6f y[k-1] + %.6f y[k-2] + %.6f u[k]\n", ...
    a1, a2, b0);


%% =========================================================
% 7. RECUPERAR LOS COEFICIENTES CONTINUOS
%
% Definimos:
%
% A = p1 + p2
% B = p1*p2
% C = p1*p2*c
%
% Modelo continuo:
%
%              C
% G(s) = ---------------
%          s^2 + A*s + B
%
%% =========================================================

D = -1/a2;

A = (a1*D - 2)/Ts;

B = (D - 1 - A*Ts)/(Ts^2);

C = b0*D/(Ts^2);

fprintf("\n===== COEFICIENTES CONTINUOS =====\n");
fprintf("A = p1 + p2 = %.10f\n", A);
fprintf("B = p1*p2   = %.10f\n", B);
fprintf("C = p1*p2*c = %.10f\n", C);


%% 8. OBTENER p1, p2 Y c

discriminante = A^2 - 4*B;

p1 = (A + sqrt(discriminante))/2;
p2 = (A - sqrt(discriminante))/2;

c = C/B;

fprintf("\n===== PARAMETROS DE LA PLANTA =====\n");
fprintf("p1 = %.10f\n", p1);
fprintf("p2 = %.10f\n", p2);
fprintf("c  = %.10f\n", c);


%% 9. CONSTRUIR G(s)

num = p1*p2*c;
den = [1, p1+p2, p1*p2];

G = tf(num, den);

fprintf("\n===== PLANTA CONTINUA ESTIMADA =====\n");
G


%% 10. VERIFICAR LA REGRESION

% Predicción de un paso usando datos medidos anteriores
y_pred = X*theta;

figure;
plot(t(3:end), Y, "DisplayName", "Medición");
hold on;
plot(t(3:end), y_pred, "--", "DisplayName", "Predicción");
grid on;
xlabel("Tiempo [s]");
ylabel("Ángulo");
title("Medición vs predicción de la regresión");
legend;